packages/api/src/command/medical/patient/update-patient.ts

```diff
@@ -22,42 +24,130 @@ export type PatientUpdateCmd = BaseUpdateCmdWithCustomer &
 
-  const patient = await updatePatientWithoutHIEs(patientUpdateWithPD, emit);
+  const patient = await updatePatientWithoutHIEs(patientUpdate, emit);
 
   const fhirPatient = toFHIR(patient);
   await upsertPatientToFHIRServer(patientUpdate.cxId, fhirPatient);
 
-  await cqCommands.patient.discover(patient, facility.id, requestId, forceCarequality);
-
-  await cwCommands.patient.update(
+  const updatedPatient = await scheduleOrRunPatientDiscovery({
     patient,
     facilityId,
-    getCqOrgIdsToDenyOnCw,
-    requestId,
-    forceCommonwell
-  );
+    rerunPdOnNewDemographics,
+    augmentDemographics,
+    forceCommonwell,
+    forceCarequality,
+  });
+
+  return updatedPatient;
+}
+
+// Needs to live somewhere else
+export async function scheduleOrRunPatientDiscovery({
+  patient,
+  facilityId,
+  rerunPdOnNewDemographics,
+  augmentDemographics,
+  forceCommonwell,
+  forceCarequality,
+  isRerunFromNewDemographics = false,
+  overrideSchedule = false,
+}: {
+  patient: Patient;
+  facilityId: string;
+  rerunPdOnNewDemographics: boolean;
+  augmentDemographics: boolean;
+  // START TODO #1572 - remove
+  forceCommonwell?: boolean;
+  forceCarequality?: boolean;
+  // END TODO #1572 - remove
+  overrideSchedule?: boolean;
+  isRerunFromNewDemographics?: boolean;
+}): Promise<Patient> {
+  let updatedPatient = patient;
+  const cqData = cqCommands.patient.getCQData(patient.data.externalData);
+
+  const discoveryStatusCq = cqData?.discoveryStatus;
+  const scheduledPdRequestCq = cqData?.scheduledPdRequest;
+
+  if (
+    discoveryStatusCq === "processing" &&
+    (!scheduledPdRequestCq || (scheduledPdRequestCq && overrideSchedule))
+  ) {
+    updatedPatient = await schedulePatientDiscovery({
+      requestId: uuidv7(),
+      patient,
+      source: MedicalDataSource.CAREQUALITY,
+      facilityId,
+      rerunPdOnNewDemographics,
+      augmentDemographics,
+      isRerunFromNewDemographics,
+    });
+  } else if (discoveryStatusCq === "processing" && scheduledPdRequestCq) {
+    // Do nothing
+  } else {
+    await cqCommands.patient.discover({
+      patient,
+      facilityId,
+      forceEnabled: forceCarequality,
+      rerunPdOnNewDemographics,
+      augmentDemographics,
+    });
+  }
+
+  const cwData = cwCommands.patient.getCWData(patient.data.externalData);
+
+  const statusCw = cwData?.status;
+  const scheduledPdRequestCw = cwData?.scheduledPdRequest;
+
+  if (
+    statusCw === "processing" &&
+    (!scheduledPdRequestCw || (scheduledPdRequestCw && overrideSchedule))
+  ) {
+    updatedPatient = await schedulePatientDiscovery({
+      requestId: uuidv7(),
+      patient,
+      source: MedicalDataSource.COMMONWELL,
+      facilityId,
+      rerunPdOnNewDemographics,
+      augmentDemographics,
+      isRerunFromNewDemographics,
+    });
+  } else if (statusCw === "processing" && scheduledPdRequestCw) {
+    // Do nothing
+  } else {
+    await cwCommands.patient.update({
```

Comment:
Not necessarily. If the case when the current PD ends, it'll be `completed` or `failed`, so this will get called. The PD status only stays in `processing` if the current PD end and there's another PD scheduled.

Reply:
But `scheduleOrRunPatientDiscovery` seems to get called before its ever marked as completed or failed

Reply:
If there's a scheduled PD run for a given HIE, the patient discovery status will not flip to `completed` but calls `discover` or `update` explicitly in the code to kick off the next run, then resets the schedule payload to allow for another run to be scheduled.

If a user creates a patient and PD runs and completes (status is `completed`) and then the user updates that patient, there will be no processing PD, so will just execute normally.

Does that make sense?

Reply:
Irrelevant now: flow has been updated to de-couple rerun and scheduling so it should be easier to see how the scheduling works. Let me know if it isn't.